merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/DirectiveTypeRegistry.java

```diff
@@ -0,0 +1,61 @@
+package gov.nasa.jpl.aerie.merlin.driver;
+
+import gov.nasa.jpl.aerie.merlin.protocol.driver.DirectiveTypeRegistrar;
+import gov.nasa.jpl.aerie.merlin.protocol.model.MissionModelFactory;
+import gov.nasa.jpl.aerie.merlin.protocol.model.TaskSpecType;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+
+public record DirectiveTypeRegistry<Registry, Model>(
+    Map<String, TaskSpecType<Model, ?, ?>> taskSpecTypes,
+    Registry registry)
+{
+  public static <Registry, Model>
+  DirectiveTypeRegistry<Registry, Model> extract(final MissionModelFactory<Registry, Model> factory) {
+    return extract(factory::buildRegistry);
+  }
+
+  public static <Registry, Model>
+  DirectiveTypeRegistry<Registry, Model> extract(final Function<DirectiveTypeRegistrar<Model>, Registry> scope) {
+    final var builder = new Builder<Model>();
+    final var registry = scope.apply(builder);
+    return new DirectiveTypeRegistry<>(builder.taskSpecTypes, registry);
+  }
+
+  public <Input, Output>
+  TaskSpecType<Model, Input, Output> lookup(final DirectiveTypeId<Input, Output> id) {
+    final TaskSpecType<Model, ?, ?> rawDirectiveType = this.taskSpecTypes.get(id.name);
+
+    @SuppressWarnings("unchecked")
+    final var directiveType = (TaskSpecType<Model, Input, Output>) rawDirectiveType;
```

Comment:
Figure it's worth a `// SAFETY:` comment here, or is it obvious from context?

Reply:
Yes, good point. I'll even add a `PRECONDITION:` note to the method itself, to call out that the method must only be invoked with IDs acquired from the same registry. (To maybe be obviated by the discussion in an earlier review comment.)